From eddfdff314c388504ca54c283ffd4b79b2a6fd4d Mon Sep 17 00:00:00 2001
From: Friedemann Kleint <Friedemann.Kleint@qt.io>
Date: Wed, 8 Oct 2025 15:19:36 +0200
Subject: [PATCH 312/553] Fix Qt Designer startup crash when restoring main
 window layout

There is another occurrence of a loop over
QMainWindowLayout::usedTabBars showing them, which causes start up
crashes of Qt Designer after a3f0ce4c0d0627ef979052b34b21e6fd2bdc3acf.

Extract the code introduced by 2c8692adaed1c3374ca01842e166af79ed3861cc
to a helper function and call it.

Task-number: QTBUG-138201
Pick-to: 6.9
Change-Id: I981abdd83725959774cde054c556cd1716c5aa7a
Reviewed-by: Volker Hilsheimer <volker.hilsheimer@qt.io>
(cherry picked from commit 080324b5eddb55bafbebb15a2b8ed71a38781259)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/widgets/widgets/qmainwindowlayout.cpp | 28 +++++++++++------------
 src/widgets/widgets/qmainwindowlayout_p.h |  1 +
 2 files changed, 15 insertions(+), 14 deletions(-)

diff --git a/src/widgets/widgets/qmainwindowlayout.cpp b/src/widgets/widgets/qmainwindowlayout.cpp
index a1a4cfc2ce0..bb116d4a3a2 100644
--- a/src/widgets/widgets/qmainwindowlayout.cpp
+++ b/src/widgets/widgets/qmainwindowlayout.cpp
@@ -1830,6 +1830,15 @@ void QMainWindowLayout::setTabPosition(Qt::DockWidgetAreas areas, QTabWidget::Ta
 QTabBar::Shape _q_tb_tabBarShapeFrom(QTabWidget::TabShape shape, QTabWidget::TabPosition position);
 #endif // QT_CONFIG(tabwidget)
 
+void QMainWindowLayout::showTabBars()
+{
+    const auto usedTabBarsCopy = usedTabBars; // list potentially modified by animations
+    for (QTabBar *tab_bar : usedTabBarsCopy) {
+        if (usedTabBars.contains(tab_bar)) // Showing a tab bar can cause another to be deleted.
+            tab_bar->show();
+    }
+}
+
 void QMainWindowLayout::updateTabBarShapes()
 {
 #if QT_CONFIG(tabwidget)
@@ -2672,11 +2681,7 @@ void QMainWindowLayout::animationFinished(QWidget *widget)
 #if QT_CONFIG(dockwidget)
         parentWidget()->update(layoutState.dockAreaLayout.separatorRegion());
 #if QT_CONFIG(tabbar)
-        const auto usedTabBarsCopy = usedTabBars; // list potentially modified by animations
-        for (QTabBar *tab_bar : usedTabBarsCopy) {
-            if (usedTabBars.contains(tab_bar)) // Showing a tab bar can cause another to be deleted.
-               tab_bar->show();
-        }
+        showTabBars();
 #endif // QT_CONFIG(tabbar)
 #endif // QT_CONFIG(dockwidget)
     }
@@ -3273,15 +3278,10 @@ bool QMainWindowLayout::restoreState(QDataStream &stream)
     savedState.deleteAllLayoutItems();
     savedState.clear();
 
-#if QT_CONFIG(dockwidget)
-    if (parentWidget()->isVisible()) {
-#if QT_CONFIG(tabbar)
-        for (QTabBar *tab_bar : std::as_const(usedTabBars))
-            tab_bar->show();
-
-#endif
-    }
-#endif // QT_CONFIG(dockwidget)
+#if QT_CONFIG(dockwidget) && QT_CONFIG(tabbar)
+    if (parentWidget()->isVisible())
+        showTabBars();
+#endif // QT_CONFIG(dockwidget)/QT_CONFIG(tabbar)
 
     return true;
 }
diff --git a/src/widgets/widgets/qmainwindowlayout_p.h b/src/widgets/widgets/qmainwindowlayout_p.h
index 464df055096..567ece0776c 100644
--- a/src/widgets/widgets/qmainwindowlayout_p.h
+++ b/src/widgets/widgets/qmainwindowlayout_p.h
@@ -643,6 +643,7 @@ private Q_SLOTS:
 #endif
 private:
 #if QT_CONFIG(tabbar)
+    void showTabBars();
     void updateTabBarShapes();
 #endif
     bool isInRestoreState = false;
-- 
2.51.2

